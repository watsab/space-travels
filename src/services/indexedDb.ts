import type { Pilot } from '@/models/Pilot';

let database: IDBDatabase;
export const getDatabase = (): Promise<IDBDatabase> => {
  return new Promise((resolve, reject) => {
    if (database) {
      return resolve(database)
    }

    let request: IDBOpenDBRequest = window.indexedDB.open('space-travels', 1)

    request.onerror = (event: Event) => {
      console.error('ERROR: Unable to open database', event);
      reject('Error');
    }

    request.onsuccess = (event: Event) => {
      database = event.target.result;
      resolve(database);
    }

    // Hook appelé à la création ou mise à jour de la bdd
    request.onupgradeneeded = (event: Event) => {
      let database: IDBDatabase = event.target.result;
      database.createObjectStore('pilots', {
        autoIncrement: true,
        keyPath: 'id',
      });
    }
  })
};


export const getPilots = (database: IDBDatabase): Promise<Pilot[]> => {
  return new Promise((resolve, reject) => {
    const transaction = database.transaction('pilots', 'readonly');
    const store = transaction.objectStore('pilots');
    const pilots: Pilot[] = [];

    store.openCursor().onsuccess = (event: Event) => {
      let cursor = event.target.result;
      if (cursor) {
        pilots.push(cursor.value as Pilot);
        cursor.continue();
      }
    }

    transaction.oncomplete = () => {
      resolve(pilots)
    };

    transaction.onerror = event => {
      reject(event)
    };
  });
};
export const getPilot = (database: IDBDatabase, pilotId: number): Promise<Pilot[]> => {
  return new Promise((resolve, reject) => {
    const transaction = database.transaction('pilots', 'readonly');
    const store = transaction.objectStore('pilots');

    let pilot: Pilot;
    store.openCursor().onsuccess = (event: Event) => {
      let cursor = event.target.result;
      if (cursor) {
        if (cursor.value.id === pilotId) {
          pilot = cursor.value;
        }
        cursor.continue();
      }
    }

    transaction.oncomplete = () => {
      resolve(pilot)
    };

    transaction.onerror = event => {
      reject(event)
    };
  });
};


export const addPilot = (database: IDBDatabase, pilot: Pilot): Promise<boolean> => {
  return new Promise((resolve, reject) => {
    const transaction = database.transaction('pilots', 'readwrite');
    const store = transaction.objectStore('pilots');

    store.put({...pilot});

    transaction.oncomplete = () => {
      console.log('Pilot successfully saved.');
      resolve(true)
    };

    transaction.onerror = event => {
      reject(event)
    };
  });
};

export const removePilot = (database: IDBDatabase, pilot: Pilot) => {
  return new Promise((resolve, reject) => {
    const transaction = database.transaction('pilots', 'readwrite');
    const store = transaction.objectStore('pilots');

    store.delete(pilot.id);
    transaction.oncomplete = () => {
      console.log('Pilot successfully saved.');
      resolve(true);
    };

    transaction.onerror = event => {
      reject(event)
    };
  });
};
