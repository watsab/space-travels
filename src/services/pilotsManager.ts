import type { Pilot } from '@/models/Pilot';
import pilotData from '@/data/pilots.json';

const getAll = (): Pilot[] => {
  return pilotData.map((pilot: Pilot): Pilot => {
    return {
      ...pilot,
      picturePath: new URL(`../assets/pilots/${pilot.picturePath}`, import.meta.url).toString()
    }
  });
}

const getById = (pilotId: number): Pilot | undefined => getAll().find(({ id }) => id === pilotId);

export {
  getAll,
  getById
}
