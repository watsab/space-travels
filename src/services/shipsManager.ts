import { usePilotsManager } from '@/composable/usePilotsManager';
import type { Ship } from '@/models/Ship';
import shipData from '@/data/ships.json';


const getAll = (): Ship[] => {
  const pilots = usePilotsManager().getAll();
  return shipData.map((ship): Ship => {
    return {
      name: ship.name,
      slug: ship.slug,
      pilot: pilots.find(({ id }) => ship.pilotId === id)
    }
  });
}

const getBySlug = (shipSlug: string): Ship | undefined => {
  getAll().find(({ slug }) => slug === shipSlug)
}

export {
  getAll,
  getBySlug,
}
