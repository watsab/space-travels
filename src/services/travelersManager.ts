import type { Traveler } from '@/models/Traveler';

const getAll = async (page = 1, limit = 5) => {
  const response = await fetch(`https://jsonplaceholder.typicode.com/users?_limit=${limit}&_page=${page}`);
  if (!response.ok) {
    throw new Error('An error occurred');
  }
  return await response.json() as Traveler[]
};

const getById = async (id: number): Promise<Traveler | null> => {
  const response = await fetch(`https://jsonplaceholder.typicode.com/users/${id}`);
  if (!response.ok) {
    if (response.status === 404) {
      return null;
    }

    throw new Error('An error occurred');
  }
  return await response.json() as Traveler;
}


export {
  getAll,
  getById,
}
