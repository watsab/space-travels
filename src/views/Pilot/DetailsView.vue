<template>
  <div>
    <router-link :to="{ name: 'pilots'}">Voir tous les pilotes</router-link>
  </div>

  <MyCard v-if="pilot">
    <template v-slot:header>
      <h2>Détails du pilote - {{ pilot.firstname }} {{ pilot.lastname }}</h2>
    </template>
    <template v-slot:body>
      <div class="pilot-information">
        <ul class="pilot-identity">
          <li><span class="text-bold">Nom:</span> {{ pilot.lastname }}</li>
          <li><span class="text-bold">Prénom:</span> {{ pilot.firstname }}</li>
        </ul>

        <img  class="pilot-picture" :src="pilot.picturePath" :alt="`${pilot.firstname} ${pilot.lastname}`">
      </div>
    </template>
  </MyCard>
</template>

<script setup lang="ts">
import { type Ref, ref, watchEffect } from 'vue';
import { useRouter } from 'vue-router';
import { useStore } from '@/store';
import type { Pilot } from '@/models/Pilot';
import MyCard from '@/components/MyCard.vue';

interface Props {
  id: number;
}

const props = defineProps<Props>();
const store = useStore();
const pilot: Ref<Pilot | null> = ref(null);
watchEffect(() => {
  store.dispatch('pilots/fetchPilots');
  pilot.value = store.getters['pilots/getById'](props.id);

  if (!pilot.value) {
    const { push } = useRouter();
    push({ name: 'notFound' });
  }
})

</script>

<style scoped>
.pilot-information {
  display: flex;
  width: 100%;
  justify-content: space-between;

  & .pilot-picture {
    max-width: 150px;
  }
}
</style>
