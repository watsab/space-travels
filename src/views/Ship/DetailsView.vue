<template>

  <div>
    <router-link :to="{ name: 'ships'}">Voir tous les vaisseaux</router-link>
  </div>


  <MyCard v-if="currentShip">
    <template v-slot:header>
      <h2>DÃ©tails du vaisseau - {{ currentShip?.name }}</h2>
    </template>
    <template v-slot:body>
      <p><span class="text-bold">Nom:</span> {{ currentShip?.name }}</p>
    </template>
  </MyCard>
</template>

<script setup lang="ts">
import { useRoute, useRouter } from 'vue-router';
import { computed,  type ComputedRef } from 'vue';
import type { Ship } from '@/models/Ship';
import { useShipsManager } from '@/composable/useShipsManager';
import MyCard from '@/components/MyCard.vue';

const route = useRoute();
const { getBySlug } = useShipsManager();
const currentShip: ComputedRef<Ship | undefined> = computed(() => {
  return getBySlug(route.params.slug as string)
})

if (!currentShip) {
  const { push } = useRouter();
  push({name: 'notFount'});
}
</script>

<style scoped>

</style>
