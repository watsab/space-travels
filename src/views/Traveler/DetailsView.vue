<template>
  <h2>DÃ©tails du voyageur - {{ traveler?.id }}</h2>
</template>

<script setup lang="ts">
import { useRouter } from 'vue-router';
import { type Ref, ref, watchEffect } from 'vue';
import type { Traveler } from '@/models/Traveler';
import { getById } from '@/services/travelersManager';


interface Props {
  id: number;
}

const props = defineProps<Props>();

const traveler: Ref<Traveler | null> = ref(null);

watchEffect(async () => {
  const { push } = useRouter();
  try {
    traveler.value = await getById(props.id);
    if (!traveler.value) {
      push({ name: 'notFound' });
    }

  } catch (error) {
    console.log(push);
    push({name: 'networkError'});
  }
})


</script>

<style scoped>

</style>
