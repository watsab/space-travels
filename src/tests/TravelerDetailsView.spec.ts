import { describe, it, expect, beforeEach, vi } from 'vitest';
import type { VueWrapper } from '@vue/test-utils';
import DetailsView from '@/views/Traveler/DetailsView.vue';
import { mount } from '@vue/test-utils';
import { getById } from '@/services/travelersManager';


const mocks = vi.hoisted(() => ({
  getById: vi.fn()
}));

vi.mock('vue-router', () => ({
  useRouter: vi.fn(() => ({
    push: vi.fn()
  })),
}));

vi.mock('@/services/travelersManager', () => ({
  getById: mocks.getById
}));


describe('Traveler/DetailsView', () => {
  let wrapper: VueWrapper<InstanceType<typeof DetailsView>>;

  beforeEach(() => {
    vi.clearAllMocks();

    wrapper = mount(DetailsView, {
      props: {
        id: 1
      }
    });
  });

  it('The rendering matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('The component calls the manager to retrieve the traveler information on id change', async () => {
    await wrapper.setProps({
      id: 123
    });

    expect(getById).toHaveBeenCalledTimes(2);
  });
})
