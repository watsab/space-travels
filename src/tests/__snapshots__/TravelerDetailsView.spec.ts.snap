// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Traveler/DetailsView > The rendering matches the snapshot 1`] = `"<h2>DÃ©tails du voyageur - </h2>"`;
